### PR TITLE
feat: Swap API - `getQuote`

### PR DESCRIPTION
**Pull request by 0xAlec**
**What changed? Why?**
- rename constants
- add `getQuote` query
- add `utils` file for formatting decimals + parameters

**Notes to reviewers**
- add docs in next PR

**How has it been tested?**

```
const eth: Token = {
  name: 'ETH',
  address: '',
  symbol: 'ETH',
  decimals: 18,
  image: 'https://wallet-api-production.s3.amazonaws.com/uploads/tokens/eth_288.png',
  chainId: 8453,
};

const degen: Token = {
  name: 'DEGEN',
  address: '0x4ed4e862860bed51a9570b96d89af5e1b0efefed',
  symbol: 'DEGEN',
  decimals: 18,
  image:
    'https://d3r81g40ycuhqg.cloudfront.net/wallet/wais/3b/bf/3bbf118b5e6dc2f9e7fc607a6e7526647b4ba8f0bea87125f971446d57b296d2-MDNmNjY0MmEtNGFiZi00N2I0LWIwMTItMDUyMzg2ZDZhMWNm',
  chainId: 8453,
};

// 1 ETH
const resp = await getQuote({ from: eth, to: degen, amount: '1' });
```

response:
```
{
  from: {
    address: '',
    chainId: 8453,
    decimals: 18,
    image: 'https://wallet-api-production.s3.amazonaws.com/uploads/tokens/eth_288.png',
    name: 'ETH',
    symbol: 'ETH'
  },
  to: {
    address: '0x4ed4e862860bed51a9570b96d89af5e1b0efefed',
    chainId: 8453,
    decimals: 18,
    image: 'https://d3r81g40ycuhqg.cloudfront.net/wallet/wais/3b/bf/3bbf118b5e6dc2f9e7fc607a6e7526647b4ba8f0bea87125f971446d57b296d2-MDNmNjY0MmEtNGFiZi00N2I0LWIwMTItMDUyMzg2ZDZhMWNm',
    name: 'DEGEN',
    symbol: 'DEGEN'
  },
  fromAmount: '1000000000000000000',
  toAmount: '185900522222140241646000',
  amountReference: 'from',
  priceImpact: '-0.07',
  chainId: 8453,
  highPriceImpact: false,
  slippage: '3'
}
```

### CODE REVIEW

File: src/swap/core/getQuote.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+export async function getQuote(params: GetQuoteParams): Promise<GetQuoteResponse> {
+  // Default parameters
+  const defaultParams = {
+    amountReference: 'from',
+    amountInDecimals: false,
+  };
+
+  params = { ...defaultParams, ...params };
+
+  const apiParams = getParamsForToken(params);
```
Comment (kyhyco) [inline src/swap/core/getQuote.ts:24]: Generally, you don't want to mutate the function argument

```suggestion
  const apiParams = getParamsForToken({ ...defaultParams, ...params });
```

File: src/swap/core/utils.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import type { GetQuoteParams, GetQuoteAPIParams } from '../types';
+
+/**
+ * Converts parameters with `Token` to ones with address. Additionally adds default values for optional request fields.
+ */
+export function getParamsForToken(params: GetQuoteParams): GetQuoteAPIParams {
```
Comment (Zizzamia) [inline src/swap/core/utils.ts:6]: In the next PR let's split each utils in their own file.

File: src/swap/types.ts [hunk 3]
```diff
@@ -3,47 +3,82 @@ import { Token } from '../token/types';
...
+  tx: Transaction; // The trade transaction
 };
 
-export type TrendingToken = Token & {
-  numOfBuys: number;
-  numOfSells: number;
+export type Transaction = {
+  data: string; // The transaction data
+  from: string; // The sender address
+  gas: string; // The gas limit
```
Comment (Zizzamia) [inline src/swap/types.ts:80]: Open qustion, should this be `gasLimit` as we have `gasPrice`?

Comment (Zizzamia) [reply src/swap/types.ts:80]: Or this is just a standard Viem and other have already? cc @0xAlec @cpcramer @kyhyco

Comment (0xAlec) [reply src/swap/types.ts:80]: will iterate on this once I integrate `getTrade` - this type is unused for `getQuote`

File: src/swap/types.ts [hunk 4]
```diff
@@ -3,47 +3,82 @@ import { Token } from '../token/types';
...
-  highPriceImpact: boolean;
-  priceImpact: string;
-  slippage: string;
-  toAmount: string;
-  toAsset: Token;
-  warning?: QuoteWarning;
+  amountReference: string; // The reference amount for the quote
+  from: Token; // The source token for the swap
+  fromAmount: string; // The amount of the source token
+  highPriceImpact: boolean; // Whether the price impact is high
```
Comment (Zizzamia) [inline src/swap/types.ts:45]: `highPriceImpact` sounds odd as `boolean`. I will expect this to have some prefix like `is` or `has`.

cc @0xAlec @cpcramer @kyhyco

File: src/swap/types.ts [hunk 5]
```diff
@@ -3,47 +3,82 @@ import { Token } from '../token/types';
...
+};
+
+/**
+ * Note: exported as public Type
+ */
+export type GetQuoteParams = {
+  from: Token; // The source token for the swap
+  to: Token; // The destination token for the swap
+  amount: string; // The amount to be swapped
+  amountReference?: string; // The reference amount for the swap
```
Comment (Zizzamia) [inline src/swap/types.ts:29]: What's a `reference`?

File: src/swap/types.ts [hunk 6]
```diff
@@ -3,47 +3,82 @@ import { Token } from '../token/types';
...
+  amountInDecimals?: boolean; // Whether the amount is in decimals
```
Comment (Zizzamia) [inline src/swap/types.ts:30]: For `boolean` we should always use some prefix like `is` or `has`

File: src/swap/types.ts (hunk 4 above)
Comment (0xAlec) [reply src/swap/types.ts:45]: I think this makes sense to me and is relatively intuitive, but not against adding a prefix either

```
const swap = await getQuote(params...)

if (swap.highPriceImpact) {
// display a warning to the user
}
```

File: src/swap/types.ts (hunk 5 above)
Comment (0xAlec) [reply src/swap/types.ts:29]: whether the `amount` in the parameter is referring to the `from` token or the `to` token (e.g. whehter you are trying to swap to 1 ETH or 1 DEGEN)

Comment (Zizzamia) [reply src/swap/types.ts:29]: ok in the next PR let's make this more clear in the comments and docs.

Folks should find easy to learn how this works just by reading the types.

**Review by Zizzamia — APPROVED** [review]

**Pull request merged**